src/ui/win/win_ui.cpp

```diff
@@ -0,0 +1,257 @@
+                HDC handleToDisplayDevice = BeginPaint( window, &paintStructure );
+
+                const WindowsUi::UiWindowWinInfo info( reinterpret_cast<UiWindowWin*>(GetWindowLongPtr( window, GWLP_USERDATA )) );
+                if ( info.valid() ) { // make sure that we draw a valid image
+                    RECT clientRoi;
+                    GetClientRect( window, &clientRoi );
+                    const DWORD result = StretchDIBits( handleToDisplayDevice, 0, 0, clientRoi.right, clientRoi.bottom,
+                                                        0, 0, info.image().width(), info.image().height(),
+                                                        info.image().data(), info.bitmapInfo(), DIB_RGB_COLORS, SRCCOPY );
+                    if ( result != info.image().height() )
+                        DebugBreak(); // drawing failed
+
+                    info.paint( handleToDisplayDevice, clientRoi );
+                }
+                EndPaint( window, &paintStructure );
+            }
+            break;
+            case WM_DESTROY:
+                PostQuitMessage( 0 ) ;
+            default:
+                return DefWindowProc( window, msg, wp, lp );
+        }
+
+        return 0;
+    }
+
+    struct UiWindowWinRegistrator
+    {
+        UiWindowWinRegistrator()
+            : registered( 0 )
+            , className( L"UiWindowWin" )
+        {
+            WNDCLASSEX wcex;
+            wcex.cbSize        = sizeof( WNDCLASSEX );
+            wcex.style         = CS_HREDRAW | CS_VREDRAW;
+            wcex.lpfnWndProc   = WindowProcedure;
+            wcex.cbClsExtra    = 0;
+            wcex.cbWndExtra    = 0;
+            wcex.hInstance     = 0;
+            wcex.hIcon         = 0;
+            wcex.hCursor       = LoadCursor( NULL, IDC_CROSS );
+            wcex.hbrBackground = reinterpret_cast<HBRUSH>(COLOR_WINDOW);
+            wcex.lpszMenuName  = NULL;
+            wcex.lpszClassName = className.data();
+            wcex.hIconSm       = 0;
+
+            registered = RegisterClassEx( &wcex );
+        }
+        WORD registered;
+        const std::wstring className;
+    };
+}
+
+UiWindowWin::UiWindowWin( const PenguinV_Image::Image & image, const std::string & title )
+    : UiWindow( PenguinV_Image::Image(), title ) // we pass an empty image into base class
+    , _bmpInfo( nullptr )
+{
+    static const UiWindowWinRegistrator registrator; // we need to register only once hence static variable
+    if ( !registrator.registered )
+        throw imageException( "Unable to create Windows API class" );
+
+    setImage( image );
+
+    const std::wstring titleWChar = std::wstring( title.begin(), title.end() );
+    _window = CreateWindowEx( 0, registrator.className.data(), titleWChar.data(), WS_OVERLAPPEDWINDOW, CW_USEDEFAULT, CW_USEDEFAULT,
+                              CW_USEDEFAULT, CW_USEDEFAULT, 0, 0, GetModuleHandle( 0 ), this ) ;
+
+    if ( _window == nullptr )
```

Comment:
`if ( _window )` to be consistent with the rest of your code.

Reply:
`_window` is not a boolean variable so comparing it with `nullptr` is totally valid.

Reply:
I agree with you. I you write `if (!_shown)` instead of `if ( shown == false)`, I was expecting the same for other checks even for pointers. Every pointer or pointer to member type can be converted to a prvalue of type `bool`.

Reply:
Yes, of course. It's just more stylistically better to see that this is a pointer. Your example (which should be `if ( !_window )` ) is equal to original code.